vyos/vyos-1x
# nat: T1083 Implement "--persistent" option to NAT rules

## Change Summary
Implemented "--persistent" option to NAT rules, options fully-random and random also implemented


## Types of changes


## Related Task(s)

* https://phabricator.vyos.net/T1083

## Component(s) name
nat


## Proposed changes
Added interface definition in conf-mode for source and destination nat for translation options, implemented nft rule modification for applying of nat translation options


## How to test
Add source or destionation nat rule with nat-flags,
```
set destination rule 100 inbound-interface 'eth0'
set destination rule 100 translation address '19.13.23.42'
set destination rule 100 translation options address-mapping 'random'
set destination rule 100 translation options port-mapping 'none'
set source rule 1000 outbound-interface 'eth0'
set source rule 1000 translation address '122.233.231.12'
set source rule 1000 translation options address-mapping 'persistent'
set source rule 1000 translation options port-mapping 'fully-random'
```
commit
check actual nft rules `nft list table nat`

```
sudo nft list table nat
table ip nat {
	chain PREROUTING {
		type nat hook prerouting priority dstnat; policy accept;
		iifname "eth0" counter packets 677 bytes 43606 dnat to 19.13.23.42 comment "DST-NAT-100"
	}

	chain INPUT {
		type nat hook input priority 100; policy accept;
	}

	chain POSTROUTING {
		type nat hook postrouting priority srcnat; policy accept;
		oifname "eth0" counter packets 48 bytes 2496 snat to 122.233.231.12 fully-random,persistent comment "SRC-NAT-1000"
	}

	chain OUTPUT {
		type nat hook output priority -100; policy accept;
	}

}
```




## Checklist:
